HLTrigger/Egamma/src/HLTEgammaGenericQuadraticFilter.cc

```diff
@@ -93,28 +135,72 @@ HLTEgammaGenericQuadraticFilter::hltFilter(edm::Event& iEvent, const edm::EventS
   edm::Handle<reco::RecoEcalCandidateIsolationMap> depMap;
   iEvent.getByToken (varToken_,depMap);
 
+  // Get rho if needed
+  edm::Handle<double> rhoHandle;
+  double rho = 0.0;
+  if (doRhoCorrection_) {
+    iEvent.getByToken(rhoToken_, rhoHandle);
+    rho = *(rhoHandle.product());
+  }
+
+  if (rho > rhoMax_)
+    rho = rhoMax_;
+  rho = rho * rhoScale_;
+
   // look at all photons, check cuts and add to filter object
   int n = 0;
+  for (unsigned int i=0; i<recoecalcands.size(); i++) {
 
-  for (auto & recoecalcand : recoecalcands) {
-
-    ref = recoecalcand;
+    ref = recoecalcands[i];
     reco::RecoEcalCandidateIsolationMap::const_iterator mapi = (*depMap).find( ref );
 
     float vali = mapi->val;
-    float energy = ref->superCluster()->energy();
     float EtaSC = ref->eta();
+
+    // Pick the right EA and do rhoCorr
+    if (doRhoCorrection_) {
+      int iEA = -1;
+      for (int cIt = absEtaLowEdges_.size() - 1; cIt > -1; cIt--) {
+        if ( std::abs(EtaSC) >= absEtaLowEdges_.at(cIt) ) {
+          iEA = cIt;
+          break;
+        }
+      }
+
+      vali = vali - (rho * effectiveAreas_.at(iEA));
+    }
+
+    float energy = ref->superCluster()->energy();
     if (useEt_) energy = energy * sin (2*atan(exp(-EtaSC)));
-    if (energy < 0.) energy=0.; /* first and second order terms assume non-negative energies */
+    if (energy < 0.) energy = 0.; /* first and second order terms assume non-negative energies */
+
+    // Pick the right cut threshold
+    double cutRegularEB_ = 9999., cutRegularEE_ = 9999.;
+    double cutOverEEB_ = 9999., cutOverEEE_ = 9999.;
+    double cutOverE2EB_ = 9999., cutOverE2EE_ = 9999.;
+
+    int iEn = -1;
+    for (int dIt = energyLowEdges_.size() - 1; dIt > -1; dIt--) {
+      if ( energy >= energyLowEdges_.at(dIt) ) {
```

Comment:
please consider replacing this loop with something like `std::lower_bound` or `std::upper_bound(energyLowEdges_.begin(), energyLowEdges_.end(), energy)`